photutils/psf.py

```diff
@@ -107,94 +118,190 @@ def evaluate(self, x, y, amplitude, x_0, y_0):
-                m = self.fitter(self, x, y, sub_array_data)
-            return m.amplitude.value
+
+        # Check input array type and dimension.
+        if np.iscomplexobj(imdata):
+            raise TypeError('Complex type not supported')
+        if imdata.ndim != 2:
+            raise ValueError('{0}-d array not supported. '
+                             'Only 2-d arrays supported.'.format(imdata.ndim))
+        if size % 2 == 0:
+            raise TypeError("Size must be odd.")
+
+        if fluxes is not None and len(fluxes) != len(positions):
+            raise TypeError("Position and flux arrays must be of equal length.")
+
+        if mask is None:
+            mask = np.isnan(imdata)
+
+        if isinstance(positions, (list, tuple)):
+            positions = np.array(positions)
+
+        if isinstance(positions, Table) or \
+            (isinstance(positions, np.ndarray) and positions.dtype.names is not None):
+            # Can do clever things like
+            # positions['x_0', 'y_0'].as_array().view((positions['x_0'].dtype, 2))
+            # but that requires  positions['x_0'].dtype is positions['y_0'].dtype
+            # better do something simple to allow type promotion if required.
+            pos = np.empty((len(positions), 2))
+            pos[:, 0] = positions['x_0']
+            pos[:, 1] = positions['y_0']
+            positions = pos
+
+        if isinstance(fluxes, (list, tuple)):
+            fluxes = np.array(fluxes)
+
+        if mode == 'mean':
+            combine = np.ma.mean
+        elif mode == 'median':
+            combine = np.ma.median
         else:
-            return 0
+            raise Exception('Invalid mode to combine prfs.')
+
+        data_internal = np.ma.array(data=imdata, mask=mask)
+        prf_model = np.ndarray(shape=(subsampling, subsampling, size, size))
+        positions_subpixel_indices = np.array([subpixel_indices(_, subsampling)
+                                               for _ in positions], dtype=np.int)
+
+        for i in range(subsampling):
+            for j in range(subsampling):
+                extracted_sub_prfs = []
+                sub_prf_indices = np.all(positions_subpixel_indices == [j, i],
+                                         axis=1)
+                positions_sub_prfs = positions[sub_prf_indices]
+                for k, position in enumerate(positions_sub_prfs):
+                    x, y = position
+                    extracted_prf = extract_array(data_internal, (size, size),
+                                                  (y, x))
+                    # Check shape to exclude incomplete PRFs at the boundaries
+                    # of the image
+                    if (extracted_prf.shape == (size, size) and
+                            np.ma.sum(extracted_prf) != 0):
+                        # Replace NaN values by mirrored value, with respect
+                        # to the prf's center
```

Comment:
I think this is a bit of a kludge to get around the problem that `NaN`s don't play well with a lot of arithmetic.  PSFs are generally at least *somewhat* symmetric, especially for space missions.  And the use case for this class is undersampled data, so just picking the mean of the surrounding points is probably also insufficient. That said, other kinds of interpolation make as much sense, this is definitely a special case.

(As above, though, this isn't my addition - I just moved it to a different place)